### PR TITLE
Prevent allow/reject reviews on merged/closed PRs

### LINKED ISSUE #30675: Merged/Closed PR should not be approvable or request changes in diff page?
### Description

Follow #29965
image

### Gitea Version

latest

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

build

### Database

None

### PR DESCRIPTION
**Pull request by kemzeb**
Resolves #30675.

### CODE REVIEW

**Comment by kemzeb** [comment]
Manually tested both approve and reject reviews on open, merged, and closed PRs and they appear to produce valid results.

Currently making this a draft PR so that I can:
- Find good places to write tests
- Catch the error object I created in the API (if we have associated API)

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
Already looking quite good :)

**Comment by kemzeb** [comment]
> Already looking quite good :)

Thanks for the review @silverwind! Just having trouble setting up a proper test for these changes since I need to use a `git.Repository` object in `SubmitReview()`  where the code expects an actual git repository to exist. I'll look to see if I can write an integration test for both closed and merged PRs 👍

**Comment by lunny** [comment]
But they should still be allowed to add comment without approve/reject. So I think the template should not be changed too much. Just disable that two buttons.

**Comment by kemzeb** [comment]
> But they should still be allowed to add comment without approve/reject. So I think the template should not be changed too much. Just disable that two buttons.

Thanks for the review! The user should be able to create a review comment with my changes. Here is an example with a closed PR:
Screenshot 2024-04-24 at 9 48 47 PM

This appears to be in-line with GitHub (though you only see a "submit review" button rather than a "comment" button). If you wish, I can go ahead and change it to just conditionally add the `disabled` attribute to the buttons.

**Comment by lunny** [comment]
> > But they should still be allowed to add comment without approve/reject. So I think the template should not be changed too much. Just disable that two buttons.
> 
> Thanks for the review! The user should be able to create a review comment with my changes. Here is an example with a closed PR: Screenshot 2024-04-24 at 9 48 47 PM
> 
> This appears to be in-line with GitHub (though you only see a "Submit Review" button rather than a "Comment" button). If you wish, I can go ahead and change it to just conditionally add the `disabled` attribute to the buttons.

I think the current design is OK for me.

**Comment by kemzeb** [comment]
Recent change adds integration tests and has the web handler just return the HTTP status rather than changing the response body.

Please take a careful eye on the test code as I could be misusing the testing API or there could be an opportunity to improve the test performance that I haven't seen.

**Comment by kemzeb** [comment]
Hmm, going to take a look at the second test again as I'm not sure if its valid

File: services/pull/review.go [hunk 1]
```diff
@@ -293,6 +310,10 @@ func SubmitReview(ctx context.Context, doer *user_model.User, gitRepo *git.Repos
 	if reviewType != issues_model.ReviewTypeApprove && reviewType != issues_model.ReviewTypeReject {
 		stale = false
 	} else {
+		if issue.IsClosed || pr.HasMerged {
```
Comment (wxiaoguang) [inline services/pull/review.go]: Just a question, in which case,  `HasMerged=true` doesn't mean `IsClosed=true`?

File: services/pull/review.go [hunk 2]
```diff
@@ -43,6 +43,23 @@ func (err ErrDismissRequestOnClosedPR) Unwrap() error {
 	return util.ErrPermissionDenied
 }
 
+// ErrSubmitReviewOnClosedPR represents an error when an user tries to submit an approve or reject review associated to a closed or merged PR.
+type ErrSubmitReviewOnClosedPR struct{}
+
+// IsErrSubmitReviewOnClosedPR checks if an error is an ErrSubmitReviewOnClosedPR.
+func IsErrSubmitReviewOnClosedPR(err error) bool {
```
Comment (wxiaoguang) [inline services/pull/review.go]: ~~Maybe we can use `errors.As` as much as possible? Many `IsErrXxx` were written for old Golang.~~

Hmm, it's also good to keep it to make code easier to search. Either is fine to me.

File: services/pull/review.go (hunk 1 above)
Comment (kemzeb) [reply services/pull/review.go]: Initially I was confused what you meant here but now I understand. `IsClosed` will always be set if the PR was merged, so it doesn't make sense to include `pr.HasMerged`. Just to make sure I took a [gander](https://github.com/go-gitea/gitea/blob/27861d711b6284ccc774f974d8a5813ca2c488eb/models/issues/pull.go#L502) at `PullRequest.SetMerged()` and this does appear to be the case after reading this. I'll make the necessary changes!

Comment (kemzeb) [reply services/pull/review.go]: Fixed

File: templates/repo/diff/new_review.tmpl [hunk 3]
```diff
@@ -30,20 +30,25 @@
...
 				<div class="divider"></div>
 				{{$showSelfTooltip := (and $.IsSigned ($.Issue.IsPoster $.SignedUser.ID))}}
-				{{if $showSelfTooltip}}
-					<span class="tw-inline-block" data-tooltip-content="{{ctx.Locale.Tr "repo.diff.review.self_approve"}}">
-						<button type="submit" name="type" value="approve" disabled class="ui submit primary tiny button btn-submit">{{ctx.Locale.Tr "repo.diff.review.approve"}}</button>
-					</span>
-				{{else}}
-					<button type="submit" name="type" value="approve" class="ui submit primary tiny button btn-submit">{{ctx.Locale.Tr "repo.diff.review.approve"}}</button>
+				{{$isOpen := (not (or $.Issue.IsClosed (and $.Issue.IsPull $.Issue.PullRequest.HasMerged)))}}
+				{{if $isOpen}}
```
Comment (wxiaoguang) [inline templates/repo/diff/new_review.tmpl]: How about also simplifying here? `if not $.Issue.IsClosed`

File: services/pull/review.go (hunk 2 above)
Comment (kemzeb) [reply services/pull/review.go]: Gotcha. After looking at the `errors` package I decided to use `errors.Is` and just turn `ErrSubmitReviewOnClosedPR` to a globally-accessible error object as this seems to be a similar pattern in other areas of the codebase and on top of this I don't need to do anything complex when handling this error.

**Review by wxiaoguang — APPROVED** [review]

File: templates/repo/diff/new_review.tmpl (hunk 3 above)
Comment (kemzeb) [reply templates/repo/diff/new_review.tmpl]: Good catch, will do now

File: services/pull/review.go (hunk 2 above)
Comment (wxiaoguang) [reply services/pull/review.go]: Golang has a bad error system ....

`Is` checkes "value", eg: `var MyErr = errors.New("...")`

`As` checks the "type", eg: `var MyErr = MyErrStruct{}`

Not easy to use, so I edited my suggestion, either is fine to me.

**Comment by kemzeb** [comment]
Yikes not sure why I decided to move the `IsClosed` check to the top as this has users unable to comment on the close/merged PR. Will make the necessary change now

File: templates/repo/diff/new_review.tmpl (hunk 3 above)
Comment (kemzeb) [reply templates/repo/diff/new_review.tmpl]: Fixed

**Comment by kemzeb** [comment]
Just to make sure, I manually tested both the API and web route with closed/merged PRs and everything appears fine from what I can see.

**Pull request merged**